src/vec.h

```diff
@@ -214,12 +259,56 @@ class Vec : public VecView<T> {
     capacity_ = this->size_;
   }
 
-  size_t capacity() const { return capacity_; }
+  inline size_t capacity() const { return capacity_; }
+
+  void MakeUnique() {
+    if constexpr (shared) {
+      if (count_->load() > 1) {
+        *this = Vec<T, true>(this->view());
+      }
+    }
+  }
 
  private:
   size_t capacity_ = 0;
 
+#if __cplusplus >= 202002L
+  [[no_unique_address]]
+#endif
+  std::conditional_t<shared, std::atomic<int>*, Empty> count_;
+
+  inline void AssertUnique() const {
+    if constexpr (shared) {
+      ASSERT(count_->load() == 1, logicErr("can only mutate unique vector"));
+    }
+  }
+
+  inline void dealloc() {
+    if constexpr (shared) {
+      if (count_ == nullptr || count_->fetch_sub(1) > 1) return;
+      delete count_;
+      this->count_ = nullptr;
+    }
+    if (this->ptr_ != nullptr) {
+      Vec<T, false>::free_async(this->ptr_, this->capacity_);
+    }
+    this->ptr_ = nullptr;
+    this->size_ = 0;
+    this->capacity_ = 0;
+  }
+
+  template <typename V>
+  inline void moveContent(V&& other) {
+    this->ptr_ = other.ptr_;
+    this->size_ = other.size_;
+    this->capacity_ = other.capacity_;
+    other.ptr_ = nullptr;
+    other.size_ = 0;
+    other.capacity_ = 0;
+  }
+
   static_assert(std::is_trivially_destructible<T>::value);
+  friend Vec<T, !shared>;
```

Comment:
This is required so we can access private fields between shared and non-shared vectors.